Structured concurrency revisions for third review (#1368)
apple/swift-evolution: proposals/0304-structured-concurrency.md

* Status: **Active Review (March 31 - April 16 2021)**
* Implementation: Available in [recent `main` snapshots](https://swift.org/download/#snapshots) behind the flag `-Xfrontend -enable-experimental-concurrency`

## Table of contents

* [Introduction](#introduction)
* [Motivation](#motivation)
* [Structured concurrency](#structured-concurrency-1)
   * [Tasks](#tasks)
   * [Child tasks](#child-tasks)

>    * [Partial tasks](#partial-tasks)
>    * [Executors](#executors)

Reviewer: Should "partial task" be renamed to "job" in these two sections?

```suggestion
   * [Jobs](#jobs)
   * [Executors](#executors)
```

Author: Hmm. That's the term we use in the implementation nowadays. Sure, that makes sense. Thanks!

   * [Task priorities](#task-priorities)
   * [Priority Escalation](#priority-escalation)
* [Proposed solution](#proposed-solution)
   * [Task groups and child tasks](#task-groups-and-child-tasks)
   * [Asynchronous programs](#asynchronous-programs)
   * [Cancellation](#cancellation)

>    * [Un-structured tasks](#un-structured-tasks)

Reviewer: ```suggestion
   * [Unstructured tasks](#unstructured-tasks)
```

      * [Context inheritance](#context-inheritance)
      * [Detached tasks](#detached-tasks)
* [Detailed design](#detailed-design)
   * [Task API](#task-api)
      * [The Task type](#the-task-type)
      * [UnsafeCurrentTask type](#unsafecurrenttask-type)
      * [Task priorities](#task-priorities-1)

>       * [Un-structured tasks](#un-structured-tasks-1)
>             * [Priority propagation](#priority-propagation)
>             * [Actor context propagation](#actor-context-propagation)
>             * [Implicit "self" in async {} closures](#implicit-self-in-async--closures)

Reviewer: ```suggestion
      * [Unstructured tasks](#unstructured-tasks-1)
         * [Priority propagation](#priority-propagation)
         * [Actor context propagation](#actor-context-propagation)
         * [Implicit "self"](#implicit-self)
```

  
  guard !Task.isCancelled else { 
    print("Cancelled mid-way through chopping of \(vegetable)!")
    throw CancellationError() 
  } 
  // chop some more, chop chop chop ...
}
```

Note also that no information is passed to the task about why it was cancelled.  A task may be cancelled for many reasons, and additional reasons may accrue after the initial cancellation (for example, if the task fails to immediately exit, it may pass a deadline). The goal of cancellation is to allow tasks to be cancelled in a lightweight way, not to be a secondary method of inter-task communication.

> ### Un-structured tasks

Reviewer: ```suggestion
### Unstructured tasks
```

So far all types of tasks we discussed were child-tasks and respected the primary rule of structured concurrency: that a *child task* cannot live longer than *parent task* (or scope) in which it was created. This is both true for task groups as well as the `async let` proposal which is being reviewed in parallel.

> Sometimes however, these rigid rules end up being too restrictive. We might need to create new tasks whose lifetime is not bound to the creating task, for example in order to fire-and-forget some operation or to initiate asynchronous work from synchronous code. Un-structured tasks are not able to utilize some of the optimization techniques wrt. allocation and metadata propagation as child-tasks are, however they remain a very important building block especially for more free-form usages and integration with legacy APIs.

Reviewer: ```suggestion
Sometimes however, these rigid rules end up being too restrictive. We might need to create new tasks whose lifetime is not bound to the creating task, for example in order to fire-and-forget some operation or to initiate asynchronous work from synchronous code. Unstructured tasks are not able to utilize some of the optimization techniques wrt. allocation and metadata propagation as child-tasks are, however they remain a very important building block especially for more free-form usages and integration with legacy APIs.
```

> All un-structured tasks are represented by a task handle, which can be used to retrieve the value (or thrown error) produced by the task, cancel the task, or perform queries of the task's status. A new task can be created with the `Task { ... }` initializer. For example:

Reviewer: ```suggestion
All unstructured tasks are represented by a task handle, which can be used to retrieve the value (or thrown error) produced by the task, cancel the task, or perform queries of the task's status. A new task can be created with the `Task { ... }` initializer. For example:
```

```swift
let dinnerHandle = Task {
  try await makeDinner()
}
```

The initializer creates a new task that begins executing the provided closure. That new task is represented by the constructed task handle (in this case, `Task<Meal, Error>`) referencing the newly-launched task. Task handles can be used to await the result of the task, e.g.,

```swift
let dinner = try await dinnerHandle.value
```

Tasks run to completion even if there are no remaining uses of their task handle, so it is not necessary to retain the task handle or observe its value for the task to complete. However, the task handle can be used to explicitly cancel the operation, e.g.,

```swift
dinnerHandle.cancel()
```

#### Context inheritance

> Un-structured tasks created with the `Task` initializer inherit important metadata information from the context in which it is created, including priority, task-local values, and actor isolation. 

Reviewer: ```suggestion
Unstructured tasks created with the `Task` initializer inherit important metadata information from the context in which it is created, including priority, task-local values, and actor isolation. 
```

If called from the context of an existing task:

- inherit the priority of the current task the _synchronous_ function is executing on

> - inherit all task-local values by copying them to the new un-structured task

Reviewer: ```suggestion
- inherit all task-local values by copying them to the new unstructured task
```

- if executed within the scope of a specific actor function:
  - inherit the actor's execution context and run the task on its executor, rather than the global concurrent one,
  - the closure passed to `Task {}` becomes actor-isolated to that actor, allowing access to the actor-isolated state, including mutable properties and non-sendable values.

If called from a context that is _not_ running inside a task:

- consult the runtime and infer the best possible priority to use (e.g. by asking for current thread priority),
- even though there is no `Task` to inherit task-local values from, check the fallback mechanism for any task-locals stored for the current synchronous context (this is discussed in depth in the SE-Thread Local Values proposal)
- execute on the global concurrent executor and be non-isolated with regard to any actor.

#### Detached tasks

> A *detached task* is an un-structured task that is independent of the context in which it is created, meaning that it does not inherit priority, task-local values, or the actor context. A new detached task can be created with the  `Task.detached` function:

Reviewer: ```suggestion
A *detached task* is an unstructured task that is independent of the context in which it is created, meaning that it does not inherit priority, task-local values, or the actor context. A new detached task can be created with the  `Task.detached` function:
```

extension TaskPriority {
  /// The task was initiated by the user and prevents the user from actively using
  /// your app.
  /// 
  /// Alias for `TaskPriority.high`.
  static var userInitiated: TaskPriority { ... }
  
  /// Priority for a utility function that the user does not track actively.
  /// 
  /// Alias for `TaskPriority.low`
  static var utility: TaskPriority { ... }
        
}

extension Task where Success == Never, Failure == Never { 
  /// Returns the `current` task's priority.
  /// 
  /// When called from a context with no `Task` available, will return the best 
  /// approximation of the current thread's priority, e.g. userInitiated for 
  /// the "main thread" or default if no specific priority can be detected. 
  static var currentPriority: TaskPriority { ... }
}
```

The `priority` operation queries the priority of the task.

The `currentPriority` operation queries the priority of the currently-executing task. Task priorities are set on task creation (e.g., `Task.detached` or `TaskGroup.async`) and can be escalated later, e.g., if a higher-priority task waits on the task handle of a lower-priority task.

> #### Un-structured tasks

Reviewer: ```suggestion
#### Unstructured tasks
```

Unstructured tasks can be created using the `Task` initializer:

```swift
extension Task where Failure == Never {
  @discardableResult
  init(
    priority: TaskPriority? = nil,
    operation: @Sendable @escaping () async -> Success
   )
}

extension Task {
  @discardableResult
  init(
    priority: TaskPriority? = nil,
    operation: @Sendable @escaping () async throws -> Success
  )
}
```

The initializers are marked with `@discardableResult` because the task itself will immediately execute the operation and run to completion when the handle is unused. This is a fairly common use case for fire-and-forget asynchronous operations.  

By default, the new task will be initially scheduled on the default global concurrent executor. Once custom executors are introduced in another proposal, these will be able to take an executor parameter to determine on which executor to schedule the new task instead.

> ###### Priority propagation

Reviewer: ```suggestion
##### Priority propagation
```

The `Task` initializer propagates priority from the point where it is called to the detached task that it creates:

1. If the synchronous code is running on behalf of a task (i.e., `withUnsafeCurrentTask` provides a non-`nil` task), use the priority of that task;
2. If the synchronous code is running on behalf of the "UI" thread, use `.userInitiated`; otherwise
3. Query the system to determine the priority of the currently-executing thread and use that.

The implementation will also propagate any other important OS-specific information from the synchronous code into the asynchronous task.

> ###### Actor context propagation

Reviewer: ```suggestion
##### Actor context propagation
```

A closure passed to the `Task` initializer will implicitly inherit the actor execution context and isolation of the context in which the closure is formed. For example:

```swift
func notOnActor(_: @Sendable () async -> Void) { }

actor A {
  func f() {
    notOnActor {
      await g() // must call g asynchronously, because it's a @Sendable closure
    }
    Task {
      g() // okay to call g synchronously, even though it's @Sendable
    }
  }
  
  func g() { }
}
```

In a sense, the `Task` initializer counteracts the normal influence of `@Sendable` on a closure within an actor. Specifically, [SE-0306](https://github.com/apple/swift-evolution/blob/main/proposals/0306-actors.md#closures) states that `@Sendable` closure are not actor-isolated:

> Actors prevent this data race by specifying that a `@Sendable` closure is always non-isolated. 

Such semantics, where the closure is both `@Sendable` and actor-isolated, are only possible because the closure is also `async`. Effectively, when the closure is called, it will immediately "hop" over to the actor's context so that it runs within the actor.

> ###### Implicit "self" in `async {}` closures

Reviewer: ```suggestion
##### Implicit "self"
```

      }
      urlSessionTask?.resume()
    }
  }, onCancel: {
    urlSessionTask?.cancel() // runs immediately when cancelled
  }
}
```

#### Voluntary Suspension

For long-running operations, say performing many computations in a tight loop
without natural suspend points, it might be beneficial to occasionally check in if the task should perhaps suspend and offer a chance for other tasks to proceed (e.g. if all are executing on a shared, limited-concurrency pool). For this use case, `Task` includes a `yield()` operation, which is a way to explicitly suspend the current task and give other tasks a chance to run for a while. 

```swift
extension Task where Success == Never, Failure == Never {
  static func yield() async { ... }
}
```

We also offer an asynchronous sleep function, which accepts the number of nanoseconds to suspend for:

```swift
extension Task where Success == Never, Failure == Never {
  public static func sleep(_ duration: UInt64) async { ... }
}
```

> The sleep function accepts a plain integer as nanoseconds to sleep for which mirrors known top-level functions performing the same action in the synchronous world. Because use-sites look quite explicit in the way they have to previous this call with an `await` keyword (`await Task.sleep(nanos)`), we prefer to use the well-known `sleep` word rather than introduce new words for this functionality.

Reviewer: [`Darwin.sleep(_:)`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sleep.html) has a `seconds: CUnsignedInt` parameter. Could the proposal specify which other known top-level functions accept nanoseconds?

Using `public static func sleep(forTimeInterval seconds: Double) async` would fit in with the [existing](https://developer.apple.com/documentation/foundation/thread/1413673-sleep) Foundation API.

Typo: "previous" → "prefix"

```suggestion
The sleep function accepts a plain integer as nanoseconds to sleep for which mirrors known top-level functions performing the same action in the synchronous world. Because use-sites look quite explicit in the way they have to prefix this call with an `await` keyword (`await Task.sleep(nanos)`), we prefer to use the well-known `sleep` word rather than introduce new words for this functionality.
```

Author: Let's fix the typo here and discuss the `sleep` API as part of the next review.
